inc/WpdbStorage/DataSourceCrud.php

```
@@ -36,8 +36,33 @@ public static function get_config_by_uuid( string $uuid ): array|WP_Error {
 		return new WP_Error( 'data_source_not_found', __( 'Data source not found', 'remote-data-blocks' ), [ 'status' => 404 ] );
 	}
 
-	public static function get_configs(): array {
-		return self::get_all_configs();
+	public static function get_configs(): array|WP_Error {
```

@alecgeatches / @chriszarate - This is the fix I'm proposing for what you two raised.

The logic I had in my head was that, this should follow the same logic as when you request an item that's disabled or soft deleted. It wouldn't be returned back under any circumstance. What that means in practice for us is,

- Get the configs
- Inflate each one against the list of known classes that we have
- If it doesn't inflate, then it means that's not valid
- Only give back all the valid configs

I have added one extra step here which is removal of unsupported data sources. This is set to false by default, as it felt like a destructive operation that we don't want to enable it yet. If it's set to true, it'll remove all the invalid configs and cleanup the DB. I also have to test it after we have enabled support for data sources that could be injected. I don't want to have them interfere and cause bad consequences. I'm happy to remove this if you two prefer.

What I think this doesn't do is add the ability to mark a data source as disabled, and then be able to re-enable it. That is a different feature imo, for when a data source is supported and one chooses to enable/ disable it. This would be when a data source isn't even supported in the first place.

---

Note: Previously, we never inflated the configs when fetching all of em. So this is a good addition to have in place.

---

> The logic I had in my head was that, this should follow the same logic as when you request an item that's disabled or soft deleted. It wouldn't be returned back under any circumstance. What that means in practice for us is,
>
> - Get the configs
> - Inflate each one against the list of known classes that we have
> - If it doesn't inflate, then it means that's not valid
> - Only give back all the valid configs

I don't agree with this from a UX / DevEx perspective. I think an invalid data source should be surfaced in the data sources list, with a clear UI indicator that it is disabled because it invalid. Ideally, we also surface any associated error messages. It doesn't have to be accomplished via this class method, but it needs to be possible.

Silently removing invalid data sources will lead to confusion. And we definitely don't want to auto-remove them from the persisted storage (option).

> What I think this doesn't do is add the ability to mark a data source as disabled, and then be able to re-enable it.

This has not been requested AFAIK.

---

Agree with Chris above! I think we want to show a disconnected/unavailable source for a couple of reasons:

- If a data source type has been removed, this is a good place to communicate that something is messed up. Silently ignoring makes it trickier to find
- If there's an unexpected configuration issue, we don't want to drop data sources accidentally

---

I get what you two are saying - just show it as disabled with the option to delete it (via the UI that is) and view any error messages necessary for this. I had been approaching this from a slightly different mindset, which was to think about all the possible cases where a data source could be disabled. I saw 4 possible cases:

- The data source has auth or other config problems. An example would if the client ID was wrong, or the auth type has been deprecated and this is an old data source.
- The data source has failed to migrate, and there is an error in this data source.
- The data source has been manually disabled.
- The data source's class itself is no longer supported. This is the case we are solving right now.

To me, that 4 case stood out as an odd ball compared to the other 3 cases. This is because, to me this is like bad data in that there is no reason to show it in the UI save to delete it. The user cannot do anything with that info, save for acking it or delete it. In addition, it does add a bit of barrier if you leave our platform that any data source which we don't support will not be shown in the data sources list.

Hence, I did it the way that I did it. I did add the auto-removal code in there but disabled it for the exact reason you said. I was 50/50 on it, more leaning towards no than yes tbh.

All this to say, that I wanted to give some more background into why I did it the way I did. What I'll do next is to try and disable this data source in the way suggested above. I'm hoping to do it in a way that other data sources can take advantage of it as well, for the other use cases I outlined above. If it's proving too challenging, I'll stick to just the one use case.

---

I've modified the code to now do the following:

- Set the enabled flag on each data source. This will be true for classes we support, and false otherwise.
- The settings UI will use the enabled flag to limit operations on disabled items to just copy and delete.

This will also leave the capacity to surface error messages within the config, if we wanted. It'll also allow us to add a filter or some other mechanism to allow third party data sources to have this flag be set too. So those can re-use this logic as well.

Hopefully I have covered my bases on this.